File: pulsar-mapr/src/main/java/ai/platon/pulsar/jobs/common/FetchEntry.java
package ai.platon.pulsar.jobs.common;

import ai.platon.pulsar.persist.WebPage;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;

public class FetchEntry extends Configured {
    private String reservedUrl;
    private WebPage page;

    public FetchEntry() {}

    public FetchEntry(Configuration conf, String reservedUrl, WebPage page) {
        super(conf);
        this.reservedUrl = reservedUrl;
        this.page = page;
    }

    public String getReservedUrl() {
        return reservedUrl;
    }

    public WebPage getWebPage() {
        return page;
    }

    @Override
    public String toString() {
        return "<" + reservedUrl + ", " + page + ">";
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/common/FetchEntry.kt
package ai.platon.pulsar.crawl.common

import ai.platon.pulsar.common.PulsarParams
import ai.platon.pulsar.common.options.LoadOptions
import ai.platon.pulsar.common.urls.NormURL
import ai.platon.pulsar.persist.WebPage

class FetchEntry(val page: WebPage, val options: LoadOptions) {

    constructor(url: String, options: LoadOptions, href: String? = null) :
            this(createPageShell(url, options, href), options)

    companion object {

        fun createPageShell(normURL: NormURL): WebPage {
            return createPageShell(normURL.spec, normURL.options, normURL.hrefSpec, normURL.referrer)
        }

        fun createPageShell(url: String, options: LoadOptions, href: String? = null, referrer: String? = null): WebPage {
            val page = WebPage.newWebPage(url, options.conf, href)
            initWebPage(page, options, href, referrer)
            return page
        }

        fun initWebPage(page: WebPage, options: LoadOptions, href: String? = null, referrer: String? = null) {
            page.also {
                it.href = href
                referrer?.let { r -> it.referrer = r }
                it.args = options.toString()
                it.fetchMode = options.fetchMode
                it.conf = options.conf
                it.maxRetries = options.nMaxRetry
                it.isResource = options.isResource

                // since LoadOptions is not visible by WebPage, we use an unsafe method to pass the load options
                it.setVar(PulsarParams.VAR_LOAD_OPTIONS, options)
            }
        }
    }
}
